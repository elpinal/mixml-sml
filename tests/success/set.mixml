link S = {type u = [LT : unit, EQ : unit, GT : unit]} with
link O = {data order = ext S.u} with

{unit MakeSet =
  link E = {module Elem =
    link T = {type t : Type} with
    {val compare : ext T.t -> ext T.t -> ext O.order}
  } with

  link X =
    link T = {type t : Type} with
    link {val empty : ext T.t} with
    link {val insert : ext E.Elem.t -> ext T.t -> ext T.t} with
    {val member : ext E.Elem.t -> ext T.t -> bool}
  seals

  link S = {type u = [Empty : unit, Node : ext X.t * ext E.Elem.t * ext X.t]} with

  link T = new [unit {data t = ext S.u}] with

  link e = {val empty = ext T.t_in (Empty () : ext S.u)} with

  link {val insert = fun (x : ext E.Elem.t) (s : ext T.t) ->
    match ext T.t_out s => {
      | Empty ()      => ext T.t_in (Node(ext e.empty, x, ext e.empty) : ext S.u)
      | Node(l, y, r) =>
          match ext O.order_out (ext E.Elem.compare x y) => {
            | LT () => ext T.t_in (Node(ext X.insert x l, y, r) : ext S.u)
            | EQ () => ext T.t_in (Node(l, x, r) : ext S.u)
            | GT () => ext T.t_in (Node(l, y, ext X.insert x r) : ext S.u)
          }
    }
  } with

  {val member = fun (x : ext E.Elem.t) (s : ext T.t) ->
    match ext T.t_out s => {
      | Empty ()      => false
      | Node(l, y, r) =>
          match ext O.order_out (ext E.Elem.compare x y) => {
            | LT () => ext X.member x l
            | EQ () => true
            | GT () => ext X.member x r
          }
    }
  }
}
