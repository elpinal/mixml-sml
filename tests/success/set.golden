{MakeSet : [forall (!18@Elem.t : Type). exist (!19 : Type). {Elem : {compare : [!18 -> !18 -> !0]-, t : [= !18]}, empty : [!19]+, insert : [!18 -> !19 -> !19]+, member : [!18 -> !19 -> bool]+, t : [= !19]}]+, order : [= !0], order_in : [[EQ : unit, GT : unit, LT : unit] -> !0]+, order_out : [!0 -> [EQ : unit, GT : unit, LT : unit]]+, u : [= [EQ : unit, GT : unit, LT : unit]]}
